azerothcore/azerothcore-wotlk
# Feat(Core/Debugging): improve crash reports

##### CHANGES PROPOSED:
- fix (enable) crash report in win
- Port all debugging scripts from [TC 3.3.5](https://github.com/TrinityCore/TrinityCore/tree/3.3.5/src/common/Debugging)

##### ISSUES ADDRESSED:
- Closes #2351 

##### TESTS PERFORMED:
- AzerothCore rev. 2d9d36718be5 2019-10-20 04:36:28 +0700 (debug branch) (Win64, RelWithDebInfo) at VS 16.3.5

##### HOW TO TEST THE CHANGES:
- Added test module https://github.com/Winfidonarleyan/mod-test-crash
- Write in-game `.test crash`
- Check core folder
For example:
![image](https://user-images.githubusercontent.com/18329778/67158783-8d0a7f80-f366-11e9-9e41-fd5892de8206.png)
- Open the folder `Crashes`
- Open last `.txt` file example: `2d9d36718be5_worldserver.exe_[20-10_18-15-10].txt`
- Check the first 51 lines
For example:

```
Revision: AzerothCore rev. 2d9d36718be5 2019-10-20 04:36:28 +0700 (debug branch) (Win64, RelWithDebInfo)
Date 20:10:2019. Time 18:15 
//=====================================================
*** Hardware ***
Processor: AMD Ryzen 5 3550H with Radeon Vega Mobile Gfx  
Number Of Processors: 8
Physical Memory: 6223524 KB (Available: 673436 KB)
Commit Charge Limit: 18282148 KB

*** Operation System ***
Windows 10 Home Edition (Version 10.0, Build 18362)

//=====================================================
Exception code: C0000005 ACCESS_VIOLATION
Fault address:  00007FF6652CE7C8 01:000000000063D7C8 D:\Core\Azeroth_Clean\worldserver.exe

Registers:
RAX:0000000000000000
RBX:0000000000000000
RCX:0000000000002002
RDX:00007FFBD281B770
RSI:000001969E4AF8B0
RDI:0000000000000000
R8: 00000196C83AC745
R9: 0000000000000000
R10:00000055EEBFEF40
R11:FFFFFFFFFFFFFFFF
R12:000001969E55D708
R13:00000055EEBFF588
R14:0000000000000000
R15:0000000000000001
CS:RIP:0033:00007FF6652CE7C8
SS:RSP:002B:00000000EEBFF0E0  RBP:EEBFF220
DS:002B  ES:002B  FS:0053  GS:002B
Flags:00010202

Call stack:
Address   Frame     Function      SourceFile
00007FF6652CE7C8  00000055EEBFF100  TestCrash_CS::HandleTestCrashCommand+18  D:\Git\azerothcore-wotlk\modules\mod-test-crash\src\TestCrash_SC.cpp line 34
00007FF664FED40E  00000055EEBFF2B0  ChatHandler::ExecuteCommandInTable+40E  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 300
00007FF664FED7C8  00000055EEBFF460  ChatHandler::ExecuteCommandInTable+7C8  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 277
00007FF664FEE690  00000055EEBFF4F0  ChatHandler::ParseCommands+F0  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 430
00007FF66523DD92  00000055EEBFF670  WorldSession::HandleMessagechatOpcode+712  D:\Git\azerothcore-wotlk\src\server\game\Handlers\ChatHandler.cpp line 297
00007FF664DD5229  00000055EEBFF7A0  WorldSession::Update+499  D:\Git\azerothcore-wotlk\src\server\game\Server\WorldSession.cpp line 323
00007FF664D44BC6  00000055EEBFF850  World::UpdateSessions+326  D:\Git\azerothcore-wotlk\src\server\game\World\World.cpp line 2636
00007FF664D43AB1  00000055EEBFF8E0  World::Update+2D1  D:\Git\azerothcore-wotlk\src\server\game\World\World.cpp line 2170
00007FF664D258B0  00000055EEBFF9A0  WorldRunnable::run+120  D:\Git\azerothcore-wotlk\src\server\worldserver\WorldThread\WorldRunnable.cpp line 49
00007FF6652942CF  00000055EEBFF9D0  std::thread::_Invoke<std::tuple<void (__cdecl*)(void *),void *>,0,1>+F  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\thread line 40
00007FFBD2750E72  00000055EEBFFA00  _beginthreadex+142
00007FFBD3987BD4  00000055EEBFFA30  BaseThreadInitThunk+14
00007FFBD4B0CED1  00000055EEBFFAB0  RtlUserThreadStart+21
```

##### KNOWN ISSUES AND TODO LIST:

- [ ] Test in Linux
- [ ] Test in Windows
- [ ] Test in Docker

##### Target branch(es):

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
